runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/ImpulseSourceFunction.java

```diff
@@ -18,30 +18,46 @@
 package org.apache.beam.runners.flink.translation.functions;
 
 import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.Iterables;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeutils.base.BooleanSerializer;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
 import org.apache.flink.streaming.api.functions.source.SourceFunction;
 
 /**
  * Source function which sends a single global impulse to a downstream operator. It may keep the
  * source alive although its work is already done. It will only shutdown when the streaming job is
  * cancelled.
  */
-public class ImpulseSourceFunction implements SourceFunction<WindowedValue<byte[]>> {
+public class ImpulseSourceFunction
+    implements SourceFunction<WindowedValue<byte[]>>, CheckpointedFunction {
 
   /** Keep source running even after it has done all the work. */
   private final boolean keepSourceAlive;
 
   /** Indicates the streaming job is running and the source can produce elements. */
   private volatile boolean running;
 
+  /** Checkpointed state which indicates whether the impulse has finished. */
+  private transient ListState<Boolean> impulseEmitted;
```

Comment:
I guess ValueState can not be here because it is not re-scale friendly?

Reply:
This source function is not keyed, so it can't make use of the `KeyedStateStore`. The `OperatorStateStore` only provides ListState and BroadcastState. 

Btw, this source is a `SourceFunction` which always runs with a parallelism of 1, so we do not even have re-partitioning of the state.